File: src/main/java/uk/gov/justice/digital/hmpps/cmd/api/config/WebClientConfiguration.java
package uk.gov.justice.digital.hmpps.cmd.api.config;

import io.netty.channel.ChannelOption;
import io.netty.handler.timeout.ReadTimeoutHandler;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.client.reactive.ClientHttpConnector;
import org.springframework.http.client.reactive.ReactorClientHttpConnector;
import org.springframework.security.oauth2.client.AuthorizedClientServiceOAuth2AuthorizedClientManager;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientManager;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientProviderBuilder;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
import org.springframework.security.oauth2.client.endpoint.DefaultClientCredentialsTokenResponseClient;
import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
import org.springframework.security.oauth2.client.web.DefaultOAuth2AuthorizedClientManager;
import org.springframework.security.oauth2.client.web.OAuth2AuthorizedClientRepository;
import org.springframework.security.oauth2.client.web.reactive.function.client.ServletOAuth2AuthorizedClientExchangeFilterFunction;
import org.springframework.web.context.annotation.RequestScope;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.netty.http.client.HttpClient;
import uk.gov.justice.digital.hmpps.cmd.api.utils.UserContext;

@Configuration
public class WebClientConfiguration {

    private final String elite2ApiRootUri;
    private final String csrRootUri;

    public WebClientConfiguration(
            @Value("${elite2api.endpoint.url}") final String eliteHealthRootUri,
            @Value("${csr.endpoint.url}") final String csrRootUri) {

        this.elite2ApiRootUri = eliteHealthRootUri;
        this.csrRootUri = csrRootUri;
    }

    @Bean
    @RequestScope
    public WebClient elite2ApiWebClient(final ClientRegistrationRepository clientRegistrationRepository,
                                     final OAuth2AuthorizedClientRepository authorizedClientRepository,
                                     final WebClient.Builder builder) {
        
        return getOAuthWebClient(authorizedClientManager(clientRegistrationRepository, authorizedClientRepository), builder, elite2ApiRootUri);
    }

    @Bean
    @RequestScope
    public WebClient csrApiWebClient(final ClientRegistrationRepository clientRegistrationRepository,
                                       final OAuth2AuthorizedClientRepository authorizedClientRepository,
                                       final WebClient.Builder builder) {
        return getOAuthWebClient(authorizedClientManager(clientRegistrationRepository, authorizedClientRepository), builder, csrRootUri);
    }

    @Bean
    public OAuth2AuthorizedClientManager authorizedClientManagerAppScope(final ClientRegistrationRepository clientRegistrationRepository,
                                                                         final OAuth2AuthorizedClientService oAuth2AuthorizedClientService) {

        final var authorizedClientProvider = OAuth2AuthorizedClientProviderBuilder.builder().clientCredentials().build();
        final var authorizedClientManager = new AuthorizedClientServiceOAuth2AuthorizedClientManager(clientRegistrationRepository, oAuth2AuthorizedClientService);

        authorizedClientManager.setAuthorizedClientProvider(authorizedClientProvider);
        return authorizedClientManager;
    }

    @Bean
    public WebClient elite2WebClientAppScope(@Qualifier(value = "authorizedClientManagerAppScope") final OAuth2AuthorizedClientManager authorizedClientManager, final WebClient.Builder builder) {
         return getOAuthWebClient(authorizedClientManager, builder, elite2ApiRootUri);
    }

    @Bean
    public WebClient csrAPIWebClientAppScope(@Qualifier(value = "authorizedClientManagerAppScope") final OAuth2AuthorizedClientManager authorizedClientManager, final WebClient.Builder builder) {
        HttpClient httpClient = HttpClient.create()
                .tcpConfiguration(tcpClient -> {
                    tcpClient = tcpClient.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 3_600_000);
                    return tcpClient;
                });
        builder.clientConnector(new ReactorClientHttpConnector(httpClient));
        return getOAuthWebClient(authorizedClientManager, builder, csrRootUri);
    }

    private WebClient getOAuthWebClient(final OAuth2AuthorizedClientManager authorizedClientManager, final WebClient.Builder builder, final String rootUri) {
        final var oauth2Client = new ServletOAuth2AuthorizedClientExchangeFilterFunction(authorizedClientManager);
        oauth2Client.setDefaultClientRegistrationId("elite2-api");

        return builder.baseUrl(rootUri)
                .apply(oauth2Client.oauth2Configuration())
                .build();
    }

    private OAuth2AuthorizedClientManager authorizedClientManager(final ClientRegistrationRepository clientRegistrationRepository,
                                                                  final OAuth2AuthorizedClientRepository authorizedClientRepository) {

        final var defaultClientCredentialsTokenResponseClient = new DefaultClientCredentialsTokenResponseClient();
        final var authentication = UserContext.INSTANCE.getAuthentication();

        defaultClientCredentialsTokenResponseClient.setRequestEntityConverter(grantRequest -> {
            final var converter = new CustomOAuth2ClientCredentialsGrantRequestEntityConverter();
            final var username = authentication.getName();
            return converter.enhanceWithUsername(grantRequest, username);
        });

        final var authorizedClientProvider =
                OAuth2AuthorizedClientProviderBuilder.builder()
                        .clientCredentials(clientCredentialsGrantBuilder -> clientCredentialsGrantBuilder.accessTokenResponseClient(defaultClientCredentialsTokenResponseClient))
                        .build();

        final var authorizedClientManager = new DefaultOAuth2AuthorizedClientManager(clientRegistrationRepository, authorizedClientRepository);

        authorizedClientManager.setAuthorizedClientProvider(authorizedClientProvider);
        return authorizedClientManager;
    }
}

File: src/main/java/uk/gov/justice/digital/hmpps/cmd/api/config/WebClientConfiguration.kt
package uk.gov.justice.digital.hmpps.cmd.api.config

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.client.SimpleClientHttpRequestFactory
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientManager
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientProviderBuilder
import org.springframework.security.oauth2.client.endpoint.RestClientClientCredentialsTokenResponseClient
import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository
import org.springframework.security.oauth2.client.web.DefaultOAuth2AuthorizedClientManager
import org.springframework.security.oauth2.client.web.OAuth2AuthorizedClientRepository
import org.springframework.security.oauth2.client.web.client.OAuth2ClientHttpRequestInterceptor
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.client.RestClient
import org.springframework.web.context.annotation.RequestScope
import uk.gov.justice.digital.hmpps.cmd.api.utils.UserContext.getAuthentication
import java.time.Duration
import kotlin.apply as kotlinApply

@Configuration
class WebClientConfiguration(
  @Value("\${api.base.url.prison-api}") private val prisonApiRootUri: String,
  @Value("\${api.base.url.csr}") private val csrRootUri: String,
  @Value("\${api.timeout:90s}") val timeout: Duration,
  @Value("\${csr.timeout}") private val csrApiTimeout: Duration,
) {

  @Bean
  @RequestScope
  fun prisonApiWebClient(
    clientRegistrationRepository: ClientRegistrationRepository,
    authorizedClientRepository: OAuth2AuthorizedClientRepository,
    builder: RestClient.Builder,
  ): RestClient = builder.authorisedRestClient(
    authorizedClientManager = authorizedClientManagerRequestScope(clientRegistrationRepository, authorizedClientRepository),
    registrationId = "prison-api",
    url = prisonApiRootUri,
    timeout = timeout,
  )

  @Bean
  @RequestScope
  fun csrApiWebClient(
    clientRegistrationRepository: ClientRegistrationRepository,
    authorizedClientRepository: OAuth2AuthorizedClientRepository,
    builder: RestClient.Builder,
  ): RestClient = builder.authorisedRestClient(
    authorizedClientManager = authorizedClientManagerRequestScope(clientRegistrationRepository, authorizedClientRepository),
    registrationId = "prison-api",
    url = csrRootUri,
    timeout = csrApiTimeout,
  )

  @Bean
  fun prisonWebClientAppScope(
    @Qualifier("authorizedClientManager") authorizedClientManager: OAuth2AuthorizedClientManager,
    builder: RestClient.Builder,
  ): RestClient = builder.authorisedRestClient(
    authorizedClientManager = authorizedClientManager,
    registrationId = "prison-api",
    url = prisonApiRootUri,
    timeout = timeout,
  )

  @Bean
  fun csrAPIWebClientAppScope(
    @Qualifier("authorizedClientManager") authorizedClientManager: OAuth2AuthorizedClientManager,
    builder: RestClient.Builder,
  ): RestClient = builder.authorisedRestClient(
    authorizedClientManager = authorizedClientManager,
    registrationId = "prison-api",
    url = csrRootUri,
    timeout = csrApiTimeout,
  )

  private fun authorizedClientManagerRequestScope(
    clientRegistrationRepository: ClientRegistrationRepository,
    authorizedClientRepository: OAuth2AuthorizedClientRepository,
  ): OAuth2AuthorizedClientManager {
    val responseClient = RestClientClientCredentialsTokenResponseClient().kotlinApply {
      addParametersConverter {
        LinkedMultiValueMap<String, String>().kotlinApply { this.add("username", getAuthentication().name) }
      }
    }
    val authorizedClientProvider = OAuth2AuthorizedClientProviderBuilder.builder()
      .clientCredentials { clientCredentialsGrantBuilder: OAuth2AuthorizedClientProviderBuilder.ClientCredentialsGrantBuilder ->
        clientCredentialsGrantBuilder.accessTokenResponseClient(
          responseClient,
        )
      }
      .build()
    val authorizedClientManager =
      DefaultOAuth2AuthorizedClientManager(clientRegistrationRepository, authorizedClientRepository)
    authorizedClientManager.setAuthorizedClientProvider(authorizedClientProvider)
    return authorizedClientManager
  }
}

private const val DEFAULT_TIMEOUT_SECONDS: Long = 30

fun RestClient.Builder.authorisedRestClient(
  authorizedClientManager: OAuth2AuthorizedClientManager,
  registrationId: String,
  url: String,
  timeout: Duration = Duration.ofSeconds(DEFAULT_TIMEOUT_SECONDS),
): RestClient = baseUrl(url)
  .requestFactory(
    SimpleClientHttpRequestFactory().kotlinApply {
      this.setConnectTimeout(timeout)
      this.setReadTimeout(timeout)
    },
  )
  .requestInterceptor(
    OAuth2ClientHttpRequestInterceptor(authorizedClientManager).kotlinApply {
      setClientRegistrationIdResolver { registrationId }
    },
  )
  .build()
